File: clients/ruby/lib/happy_wakey_client.rb
require "json"
require "net/http"
require "uri"

class HappyWakeyClient
  def initialize(base_url:, token: nil, telemetry: nil)
    @base = URI(base_url); raise ArgumentError, "HTTPS required" unless @base.scheme == "https"
    @token, @telemetry = token, telemetry
  end
  def health
    call(:health, :get, "/healthz", nil, false)
  end
  def list_alarms
    call(:list_alarms, :get, "/v1/alarms")
  end
  def create_alarm(request)
    call(:create_alarm, :post, "/v1/alarms", request)
  end
  def transition_occurrence(id, request)
    call(:transition_occurrence, :post, "/v1/occurrences/#{URI.encode_www_form_component(id)}/transitions", request)
  end
  def pull_changes(cursor: "0", limit: 100)
    call(:pull_changes, :get, "/v1/sync/pull?#{URI.encode_www_form(cursor: cursor, limit: limit)}")
  end
  def push_changes(changes)
    call(:push_changes, :post, "/v1/sync/push", changes)
  end
  private
  def call(operation, method, path, body = nil, auth = true)
    raise ArgumentError, "Shared Auth bearer token required" if auth && (@token.nil? || @token.empty?)
    uri = @base + path; request = method == :get ? Net::HTTP::Get.new(uri) : Net::HTTP::Post.new(uri)
    request["Accept"] = "application/json"; request["Authorization"] = "Bearer #{@token}" if auth
    request["Content-Type"], request.body = "application/json", JSON.generate(body) unless body.nil?
    response = Net::HTTP.start(uri.host, uri.port, use_ssl: true, open_timeout: 5, read_timeout: 10) { |http| http.request(request) }
    @telemetry&.info("happy_wakey.client.request", operation: operation, status: response.code.to_i)
    raise "Happy Wakey request failed (#{response.code})" unless response.is_a?(Net::HTTPSuccess)
    JSON.parse(response.body)
  rescue => error
    @telemetry&.error("happy_wakey.client.request", operation: operation)
    raise error
  end
end

File: clients/dart/lib/happy_wakey_client.dart
import 'dart:convert';
import 'package:oresoftware_next_loggers/oresoftware_next_loggers.dart';

typedef Transport =
    Future<Map<String, Object?>> Function(
      String method,
      Uri url,
      Map<String, String> headers,
      String? body,
    );

final class HappyWakeyClient {
  HappyWakeyClient({
    required String baseUrl,
    required this.transport,
    this.token,
    this.telemetry,
    bool allowInsecureLoopback = false,
  }) : base = Uri.parse(baseUrl) {
    final loopback = const {
      'localhost',
      '127.0.0.1',
      '::1',
    }.contains(base.host);
    if (base.scheme != 'https' && !(allowInsecureLoopback && loopback))
      throw ArgumentError('HTTPS required');
  }
  final Uri base;
  final String? token;
  final Transport transport;
  final Logger? telemetry;
  Future<Object?> health() => _call('health', 'GET', '/healthz', null, false);
  Future<Object?> listAlarms() => _call('list_alarms', 'GET', '/v1/alarms');
  Future<Object?> createAlarm(Object request) =>
      _call('create_alarm', 'POST', '/v1/alarms', request);
  Future<Object?> transitionOccurrence(String id, Object request) => _call(
    'transition_occurrence',
    'POST',
    '/v1/occurrences/${Uri.encodeComponent(id)}/transitions',
    request,
  );
  Future<Object?> pullChanges({String cursor = '0', int limit = 100}) => _call(
    'pull_changes',
    'GET',
    '/v1/sync/pull?cursor=${Uri.encodeQueryComponent(cursor)}&limit=$limit',
  );
  Future<Object?> pushChanges(List<Object?> changes) =>
      _call('push_changes', 'POST', '/v1/sync/push', changes);
  Future<Object?> _call(
    String operation,
    String method,
    String path, [
    Object? body,
    bool auth = true,
  ]) async {
    if (auth && (token == null || token!.isEmpty))
      throw StateError('Shared Auth bearer token required');
    try {
      final response = await transport(method, base.resolve(path), {
        'accept': 'application/json',
        if (auth) 'authorization': 'Bearer $token',
        if (body != null) 'content-type': 'application/json',
      }, body == null ? null : jsonEncode(body));
      await telemetry?.info('happy_wakey.client.request').addFields({
        'operation': operation,
        'status': response['status'],
      }).send();
      return response['body'];
    } catch (error) {
      await telemetry?.error('happy_wakey.client.request').addFields({
        'operation': operation,
      }).send();
      rethrow;
    }
  }
}
